### PR TITLE
WCS-OCaml version 2017-05-26.02.

### PR DESCRIPTION
**Pull request by mandel**
This is the new version of the Watson Conversation Service SDK for OCaml.

### CODE REVIEW

**Comment by gasche** [comment]
All travis builds [fail](https://travis-ci.org/ocaml/opam-repository/jobs/300319229):

```
# Error: External library "menhirlib" not found.
# -> required by "wcs-lib/jbuild (context default)"
# Hint: try: jbuilder external-lib-deps --missing -p wcs-lib @install
```

**Comment by mandel** [comment]
Thanks @gasche for looking at the PR. But I don't understand this error and I don't manage to reproduce it.
It looks that menhirlib is not found but it is part of the dependencies declared in the `opam` file and in the log it is marked as installed.

Would you have any clue on how to solve the problem?

**Comment by gasche** [comment]
I relaunched the [Travis CI build](https://travis-ci.org/ocaml/opam-repository/builds/300319225?utm_source=github_status&utm_medium=notification) a couple days ago and it failed again. In my experience it is fairly robust: it seems likely that someone would encounter this issue.

Wild guess: the error message is about a `menhirlib` package, but on my system the ocamlfind package name is `menhirLib` instead (capital L), is there a typo somewhere?

**Comment by mandel** [comment]
Thanks @gasche !
You were correct, it was the error. I was not able to reproduce because by default on OSX the file system is not case sensitive.

**Comment by gasche** [comment]
There seems to be a 4.05-only failure now:

```
### stderr ###
# [...]
#       atdgen wcs-lib/wcs_t.{ml,mli}
#       atdgen wcs-lib/wcs_j.{ml,mli}
#         rmlc wcs-lib/spel_t.rzi
#         rmlc wcs-lib/json_t.rzi
#         rmlc wcs-lib/wcs_t.rzi (exit 2)
# (cd _build/default/wcs-lib && /home/travis/.opam/4.05.0/bin/rmlc.opt -c wcs_t.mli)
# Cannot find the compiled interface file json_spel_t.rzi.
```

**Comment by mandel** [comment]
It looks that the remaining errors are not directly due to wcs-ocaml. Do we have to fix them or is it possible to accept the package as it is?

`# Error: I couldn't find 'ocaml-migrate-parsetree.driver-main'.` (https://ci.ocaml.io/log/saved/docker-build-f941c09aed0426aa3736435e6138d6ab/8031d5a93ea6e1f92a62ecc8229cc00ab140c4ca)
Do we have to add this dependency even if wcs-ocaml doesn't need it?

`[ERROR] The compilation of cohttp failed at "jbuilder build -p cohttp -j 4".` (https://travis-ci.org/ocaml/opam-repository/jobs/302727938)
Do we have to add the constraint that wcs-ocaml doesn't work with OCaml 4.06 even if the error is due to cohttp?

`[ERROR] wcs-api.2017-05-26.02 is not available because your system doesn't comply with ocaml-version >= "4.03.0".` (https://ci.ocaml.io/log/saved/docker-build-8abb1142b4de2058329dc4325954b552/d61045282bee201d95a591097f8548bb544905c6)
OCaml-CI tries to compile the wcs-api and wcs packages with OCaml 4.02 even if the opam file declare that it is available only for OCaml 4.03 and above.

**Comment by gasche** [comment]
The second and third failures are clearly not your fault (the third is a glitch, unsupported OCaml versions should not count as failures, and the second is a cohttp issue). I am less sure about the first one, though, as it actually fails when building your project.

Have you tried reproducing the failure with the version of the dependencies used in the build?

```
-> installed base-num.base
-> installed num.0
-> installed conf-m4.1
-> installed ocamlfind.1.7.3
-> installed cppo.1.4.0
-> installed gen.0.5
-> installed jbuilder.1.0+beta16
-> installed easy-format.1.3.0
-> installed biniou.1.0.6
-> installed ppx_tools.5.0+4.05.0
-> installed menhir.20171013
-> installed atd.1.12.0
-> installed rml.1.09.05
-> installed yojson.1.3.3
-> installed sedlex.1.99.3
-> installed atdgen.1.12.0
```

**Comment by avsm** [comment]
For the record, this is the failure in question:
```
#=== ERROR while compiling wcs-lib.2017-05-26.02 ==============================#
# context              2.0.0~beta5 | linux/x86_64 | ocaml-base-compiler.4.05.0 | file:///home/opam/src
# path                 ~/.opam/ocaml-base-compiler.4.05.0/.opam-switch/build/wcs-lib.2017-05-26.02
# command              ~/.opam/ocaml-base-compiler.4.05.0/bin/jbuilder build -p wcs-lib -j 1
# exit-code            1
# env-file             ~/.opam/log/wcs-lib-40-874a9e.env
# output-file          ~/.opam/log/wcs-lib-40-874a9e.out
### output ###
# Error: I couldn't find 'ocaml-migrate-parsetree.driver-main'.
# I need this library in order to use ppx rewriters.
# See the manual for details.
# Hint: opam install ocaml-migrate-parsetree
```

File: packages/wcs-lib/wcs-lib.2017-05-26.02/opam [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+homepage: "https://github.com/IBM/wcs-ocaml"
+bug-reports: "https://github.com/IBM/wcs-ocaml/issues"
+license: "Apache License 2.0"
+dev-repo: "https://github.com/IBM/wcs-ocaml.git"
+
+build: [
+  ["jbuilder" "subst" "-p" name "--name" name] {pinned}
+  ["jbuilder" "build" "-p" name "-j" jobs]
+]
+build-test: [ ["jbuilder" "runtest"] ]
```
Comment (rgrinberg) [inline packages/wcs-lib/wcs-lib.2017-05-26.02/opam]: You should leave the `"-p" name` bit for build-test as well as you only want to execute tests that are spceific to a sub package.

File: packages/wcs-lib/wcs-lib.2017-05-26.02/opam [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+depends: [
+  "ocamlfind"
+  "jbuilder"
+  "atdgen"
```
Comment (rgrinberg) [inline packages/wcs-lib/wcs-lib.2017-05-26.02/opam]: Slightly more precise here is something like:

```
"atd"
"atdgen" {build}
```
As it's very unlikely that the code generator is a runtime dependency.

File: packages/wcs-lib/wcs-lib.2017-05-26.02/opam [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+dev-repo: "https://github.com/IBM/wcs-ocaml.git"
+
+build: [
+  ["jbuilder" "subst" "-p" name "--name" name] {pinned}
+  ["jbuilder" "build" "-p" name "-j" jobs]
+]
+build-test: [ ["jbuilder" "runtest"] ]
+
+depends: [
+  "ocamlfind"
```
Comment (rgrinberg) [inline packages/wcs-lib/wcs-lib.2017-05-26.02/opam]: If you're using jbuilder, it's unlikely that ocamlfind is a direct dependency of yours. Hence it should be removed.

File: packages/wcs-lib/wcs-lib.2017-05-26.02/opam (hunk 2 above)
Comment (hcarty) [reply packages/wcs-lib/wcs-lib.2017-05-26.02/opam]: `jbuilder` should be a build dependency here too, correct?

**Comment by rgrinberg** [comment]
So you should add `ocaml-migrate-parsetree` as a dependency as well (likely a `{build}` dependency).

Make sure to reflect all these changes in your upstream repo as well. Thanks for taking the time to make a quality release with us.

**Comment by gasche** [comment]
Why should `ocaml-migrate-parsetree` be a dependency if @mandel is not aware that it is a dependency? I wonder if there is something about the tooling/dependencies that adds it as an implicit dependency.

**Comment by gasche** [comment]
Travis now passes for all OCaml versions except 4.06, where `cohttp` is blocking. That sounds good-to-merge to me. @rgrinberg, do you agree?

**Comment by mandel** [comment]
Thank you all for your feedback and advises!

**Comment by avsm** [comment]
Thanks very much for the iterations on this package! You may want to announce this on <https://discuss.ocaml.org>, where we have a Community category and an `announce` tag for this purpose.

**Pull request merged**